Fixes #19.  Create a module for detecting if the CMSHashStringSalt ke…

EvanZimmerman opened the pull request:
Created a module to detect if the CMSHashStringSalt key was found within the app settings.

petrsvihlik commented:
@ondrejsevcik will you please have a look at it? i'll be off next week. thx

ondrejsevcik commented:
Sure, I will have some spare time during the weekend ;)

ondrejsevcik commented on KInspector.Modules/Modules/Security/HashSaltForMacroSignaturesModule.cs:
```diff
@@ -0,0 +1,46 @@
+
+        public ModuleResults GetResults(IInstanceInfo instanceInfo)
+        {
+           if(ConfigurationManager.AppSettings["CMSHashStringSalt"] != null)
```
It might me an empty string. Make sure you handle this scenario as well.

ondrejsevcik replied:
It is loading current's app settings (KInspector). We need to load the right web.config. Take a look in [SecurityAppSettingsModule.cs](https://github.com/Kentico/KInspector/blob/master/KInspector.Modules/Modules/Security/SecurityAppSettingsModule.cs#L52) how it's handled there.

ondrejsevcik commented on KInspector.Modules/Modules/Security/HashSaltForMacroSignaturesModule.cs:
```diff
@@ -0,0 +1,46 @@
+                    new Version("8.2"),
+                    new Version("9.0")
+                },
+                Comment = @"Ensures that the 'CMSHashStringSalt' key is found within the web configuration file."
```
We should also add link to documentation for further information about this key. Maybe you can incorporate something of that into the description of the module?

https://docs.kentico.com/display/K9/Working+with+macro+signatures#Workingwithmacrosignatures-Configuringthehashsaltformacrosignatures

ondrejsevcik commented:
Hi @EvanZimmerman,

thank you for your contribution. It's great that you've decided to improve KInspector with this feature. I believe it will help a lot of users.

I looked at the code and commented on the issues I've found. Once you fix the issues, I will be more than happy to merge it into the master and release it to our users.

EvanZimmerman commented:
Hello @ondrejsevcik, 
Thank you for the support and walking me through this task.  This is an awesome learning experience for me and will set me up for all open source contributions I make in the future! 

I have modified my code and updated the comments to include the documentation that you provided above.  I was wondering if there was a way you could possibly guide me through creating some tests for this module (assuming my updated code is now correct).  I am confused on how to go about mocking data that is included within an web configuration file.  Thanks in advanced!

ondrejsevcik commented:
@EvanZimmerman, did you push your changes? Maybe I do something wrong, but I don't see any commits since I commented on this PR.

EvanZimmerman commented:
@ondrejsevcik There is the commit.  I wasn't going to commit any changes since I wanted to do some unit testing first.  I have no real way of testing my code since I'm a bit confused on how to setup a working instance of this application.

ondrejsevcik commented:
One way to mock ConfigurationManager is simply to not mock it at all :) 
You can create a new test web.config in the tests project and then create InstanceInfo with a path to this test web.config.

```
var instanceConfig = new InstanceConfig() 
{
 ...
   Path = "path/to/test/environment"
 }
var instanceInfo = new InstanceInfo(instanceConfig);

var moduleToTest = new ModuleToTest();
var result = moduleToTest.GetResults(instanceInfo);
...
```
or you can actually mock ConfigurationManager (see [this example](http://stackoverflow.com/questions/9486087/how-to-mock-configurationmanager-appsettings-with-moq))

Also, I would recommend you to take a look at the [existing tests](https://github.com/Kentico/KInspector/blob/master/KInspector.Tests/Modules/Security/PasswordPolicyModuleTests.cs#L69) to see how they use mocking.

ChristopherJennings commented:
I've incorporated this logic into the "Security settings in web.config" module in PR #134 as it's been awhile since this PR was updated. If you end up figuring out testing, I'd ask that you apply it to that module. In the end I think it made more sense to combine this check with the other security related checks in the web.config. Thanks for working out the logic though!
